matrix_functions.py

```diff
@@ -200,15 +203,37 @@ def matrix_eigendecomposition(
             device=A.device,
         )
 
+    # TODO: reduce redundant code when noninvertible_handling_config is generalized to all methods
+    # check epsilon is 0 when using pseudo-inverse
+    if (
+        isinstance(
+            eigendecomposition_config.noninvertible_handling_config,
+            PseudoInverseConfig,
+        )
+        and epsilon != 0.0
+    ):
+        raise ValueError(f"{epsilon=} should be 0.0 when using pseudo-inverse!")
+
+    # Add epsilon to the diagonal to help with numerical stability of the eigenvalue decomposition
+    # Only do it when add_epsilon_before_computation is True (root_inv_config must be a RegularizationConfig)
+    if getattr(eigendecomposition_config, "add_epsilon_before_computation", False):
+        dim = A.shape[0]
+        identity = torch.eye(dim, dtype=A.dtype, device=A.device)
+        A = A + epsilon * identity
```

Comment:
Nit: Could consider this:
```suggestion
        A = A.add(identity, alpha=epsilon)
```